### PR TITLE
Add Safe Mode Restart Switch

### PR DESCRIPTION
**Pull request by paulmonigatti**
# What does this implement/fix? 

Adds a new Safe Mode Restart Switch that allows remote rebooting into safe mode.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** esphome/feature-requests#1439

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1504

## Test Environment

- [x] ESP32
- [x] ESP8266

## Example entry for `config.yaml`:
```yaml
# Example config.yaml
switch:
  - platform: safe_mode
    name: restart_safemode
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by jesserockz** [comment]
Can I ask why this would be needed?

**Comment by paulmonigatti** [comment]
It's mainly to enable OTA updates in cases where a component in misbehaving and interfering with the process.

As an example, I've had issues in the past with the software UART being overwhelmed with data, which has prevented OTA from working at all. In the referenced feature request (esphome/feature-requests#1439), a similar thing has been happening with the BLE component, among others.

Much like safe mode and the restart switch itself, this PR would not be necessary if all components worked correctly, 100% of the time. However, in the meantime it is nice to have the ability to recover from these problems, especially on devices that are physically hard to access.

**Comment by jesserockz** [comment]
Ok that makes sense.

This use case does not generally happen, but your code changes render a regular restart switch un-compilable if a person does not include `ota` in their config.

Perhaps a new `platform: safe_mode` might be best that has a dependency on `ota` being in config validation.

**Comment by paulmonigatti** [comment]
> This use case does not generally happen, but your code changes render a regular restart switch un-compilable if a person does not include `ota` in their config.

Oops! I've just added a sprinkling of `#ifdef`s to fix this compilation issue, but I'm happy to implement as a new component if that's the preferred approach.

**Comment by jesserockz** [comment]
> but I'm happy to implement as a new component if that's the preferred approach.

I would prefer a new switch as it would be a cleaner and more clear (to the end user) approach.

**Comment by paulmonigatti** [comment]
> I would prefer a new switch as it would be a cleaner and more clear (to the end user) approach.

Done! Docs have also been updated to reflect this change.

File: esphome/components/safe_mode/switch/safe_mode_switch.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+namespace esphome {
+namespace safe_mode {
+
+class SafeModeSwitch : public switch_::Switch, public Component {
+ public:
+  void dump_config() override;
+  void set_ota(ota::OTAComponent *ota);
+
+ protected:
+  ota::OTAComponent *ota_{nullptr};
```
Comment (jesserockz) [inline esphome/components/safe_mode/switch/safe_mode_switch.h]: You don't need the `nullptr` and the check above `if (this->ota_ != nullptr) {` as config validation will make sure ota is there and will always call the setter.

Comment (paulmonigatti) [reply esphome/components/safe_mode/switch/safe_mode_switch.h]: Okay, I've just removed these checks

**Review by jesserockz — APPROVED** [review]

**Comment by mmakaay** [comment]
Nice one. When running into the type of use cases that this switch is for, I normally had to resolve to keep bashing the reset switch on the device to bump the safe mode counter up to the level of a safe restart. So 👍

**Pull request merged**